### PR TITLE
Ask for passphrases not only on the first console of /dev/console

### PR DESCRIPTION
**Pull request by bitstreamout**
Ask for passphrases not only on the first console of /dev/console but also on all other consoles. This does help on e.g. mainframes where often a serial console together with other consoles are used. Even rack based servers attachted to both a serial console as well as having a virtual console do sometimes miss a connected monitor.

Compare with the discussion at http://lists.freedesktop.org/archives/systemd-bugs/2014-August/003350.html and also the issue at https://github.com/systemd/systemd/issues/1027

### CODE REVIEW

**Comment by steelman** [comment]
How about squashing all these commits into one?

**Comment by bitstreamout** [comment]
I've tried that but this leads to an error:

> No changes
> Could not apply 343ba3b18b478f620bd927457ab45c4c8f43d00f... Revert "Ask for passphrases not only on the first console of /dev/console"*

... looks like the remove causes trouble in the git logic

**Comment by bitstreamout** [comment]
After _git rebase --continue_  I see a broken local repository with mixed up commits that is wrong Author and a file _src/libsystemd/sd-event/sd-event.c_ changed I've never edited nor touched.  I prefere not to push this to github.

**Comment by zonque** [comment]
Create a new branch forked from master, and cherry-pick the commits you want.

**Comment by steelman** [comment]
git rebase -i 2e396de8b2317fdd1cb2f5eb1df8dffd13bf1e80

Then remove the first two commits and squash the rest.

**Comment by bitstreamout** [comment]
Thanks, it has obviously worked

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 1]
```diff
@@ -614,8 +696,90 @@ static int parse_argv(int argc, char *argv[]) {
...
+                        exit(EXIT_FAILURE);
+                }
+
+                if (consoles[id].pid == 0) {
+                        if (prctl(PR_SET_PDEATHSIG, SIGHUP) < 0)
+                                _exit(EXIT_FAILURE);
+                        zero(sig);
+                        assert_se(sigprocmask(SIG_UNBLOCK, &oldset, NULL) == 0);
+                        assert_se(sigaction(SIGCHLD, &oldsig, NULL) == 0);
+                        return wfa_child(consoles, id);
```
Comment (dvdhrm) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: Errrr, what? We _never_ return from fork() to the calling function, except if the function is just a fork-wrapper. Furthermore, why is that needed, anyway? There's O_NONBLOCK and poll().

Comment (bitstreamout) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: The fork() is used to have child for each device (e.g. /dev/ttyS0 and /dev/tty1) which simply ask for the required password/passphrase.  If a password/passphrase is typed and entered (e.g. on /dev/ttyS0) then the code will kill the other asking child.  To get this synchronous a shared memory area is used, that is the index id is used to shift a bit for each running child.  Indeed the function wait_for_answer() returns in main() for each child for its own whereas the parent process remains in wait_for_answer() to wait on the first successful password/passphrase submission

A similar code I had written some times ago for sulogin of  upstream util-linux. It does work flawless as well I've tested this change for systemd several times successfully.

Comment (dvdhrm) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: I can see what the code does, my question rather is _why_ you chose to do it this way? I cannot see why we need one process for each tty we run on. What's wrong with standard I/O multiplexing via poll(2) and friends?

Secondly, your locking around the shared memory region is missing. I cannot see how that is supposed to work?

Comment (bitstreamout) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: As reference only https://git.kernel.org/cgit/utils/util-linux/util-linux.git/commit/?id=dfbc7c89b2e5a417a7b4dfbadc7250987f43d71b

Comment (bitstreamout) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: I had already reported this with a slightly other version, see http://lists.freedesktop.org/archives/systemd-bugs/2014-August/003350.html

Comment (dvdhrm) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: The util-linux change is racy regarding access to `usemask`, just like @keszybz reported on your previous patch.

If you want to support multiple consoled for ask-password, then I'd prefer if you implement this via poll(2) and act on all TTYs in the same process.

Comment (bitstreamout) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: You're pointing to http://lists.freedesktop.org/archives/systemd-bugs/2014-August/003364.html, arn't you? I'd like to point out that we are discussing about entering a passphrase at boot.  The current code in the request works in "first-come, first-served" mode.  In my opinion a locking as well as polling seems not to be required. As far as a password/passphrase has been entered in one of the childs all other childs are not needed anymore.   Therefore I'd like to ask what advantage you and @keszybz see for polling for the passphrase in the parent process?

Comment (dvdhrm) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: This is about coding-style. We don't fork off a bunch of processes just to perform async-IO. We do this properly, by opening file-descriptors in non-blocking mode and then running poll(2) on them.

Comment (bitstreamout) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: What exactly does `We do this properly' mean?  I know that my approach works;) Don't know if I'll find the time to rewrite the code as for passwords every terminal/device access a setsid() as well as a controlling tty (as done by acquire_terminal()) is mandatory.

Comment (dvdhrm) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: If `setsid()` is required, then please mention this in your commit message. Try to explain why you chose to do it the way you did. If this is really the case, then yes, a fork() sounds like a reasonable solution.

Anyway, if it is really necessary to make the terminal the controlling TTY, then yes, please use fork(). However, I really dislike the coding style. Please fork one process per TTY, which has a proper function that performs whatever the child does and then exits. Don't return to the caller after fork().

Then make the parent wait for the first child and abort the other children.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 2]
```diff
@@ -53,6 +59,19 @@ static enum {
...
+struct console {
+        pid_t pid;
+        char *tty;
+};
+
+static volatile unsigned long *usemask;
+static volatile sig_atomic_t sigchild;
+static void chld_handler(int sig)
+{
+        (void)sig;
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c:72]: This isn't needed. We don't warn about unused function parameters.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 3]
```diff
@@ -210,6 +229,69 @@ static int ask_password_plymouth(
...
+        unsigned int n;
+        if (!con || !num)
+                return;
+        for (n = 0; n < num; n++)
+                free(con[n].tty);
+        free(con);
+}
+
+static const char *current_dev = "/dev/console";
+static struct console* collect_consoles(unsigned int * num) {
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c:243]: The coding style should be in line with the rest of our code base. In particular, it should be `char *foo`, not `char* foo` or `char * foo`.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 4]
```diff
@@ -210,6 +229,69 @@ static int ask_password_plymouth(
...
+        FOREACH_WORD(word, len, active, state) {
+                _cleanup_free_ char *tty = NULL;
+
+                if (strneq(word, "tty0", len) &&
+                    read_one_line_file("/sys/class/tty/tty0/active", &tty) >= 0) {
+                        word = tty;
+                        len = strlen(tty);
+                }
+                con = greedy_realloc((void**)&con, &con_len, 1+(*num), sizeof(struct console));
+                if (con == NULL) {
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c:265]: We don't compare pointers with `NULL`. `!con` is sufficient.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 5]
```diff
@@ -210,6 +229,69 @@ static int ask_password_plymouth(
...
+                        return NULL;
+                }
+                if (asprintf(&con[*num].tty, "/dev/%.*s", (int)len, word) < 0) {
+                        free_consoles(con, *num);
+                        log_oom();
+                        *num = 0;
+                        return NULL;
+                }
+                con[*num].pid = 0;
+                (*num)++;
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c:276]: The parentheses are not needed.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 6]
```diff
@@ -210,6 +229,69 @@ static int ask_password_plymouth(
...
+        }
+        if (con == NULL) {
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c:278]: See above.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 7]
```diff
@@ -614,8 +696,90 @@ static int parse_argv(int argc, char *argv[]) {
         return 1;
 }
 
+static unsigned int wfa_child(const struct console * con, const unsigned int id)
+{
+        setsid();
+        release_terminal();
+        *usemask |= 1 << id;   /* shared memory area */
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: This should be `1UL`. Same for other locations.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 8]
```diff
@@ -45,6 +49,8 @@
 #include "process-util.h"
 #include "terminal-util.h"
 #include "signal-util.h"
+#include "fileio.h"
+#include "macro.h"
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c:54]: Include files should be ordered alphabetically.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 9]
```diff
@@ -210,6 +229,69 @@ static int ask_password_plymouth(
         return 0;
 }
 
+static void free_consoles(struct console *con, const unsigned int num) {
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c:233]: The `const` for `unsigned int num` can be dropped.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 10]
```diff
@@ -210,6 +229,69 @@ static int ask_password_plymouth(
...
+static const char *current_dev = "/dev/console";
+static struct console* collect_consoles(unsigned int * num) {
+        _cleanup_free_ char *active = NULL;
+        const char *word, *state;
+        struct console *con = NULL;
+        size_t con_len = 0, len;
+        int ret;
+
+        assert(num);
+        assert(*num == 0);
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c:251]: Why this assertion?

We generally have the rule that we don't touch the pointers to return values unless we exit with success (0). Hence, you should keep the count variable local and only set `*num` at the very end of this function.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 11]
```diff
@@ -627,11 +791,27 @@ int main(int argc, char *argv[]) {
...
+         * Use this shared memory area to be able to synchronize the
+         * workers asking for password with the main process.
+         * This allows to continue if one of the consoles had been
+         * used as afterwards the remaining asking processes will
+         * be terminated.  The wait_for_terminate() does not help
+         * for this use case.
+         */
+        usemask = mmap(NULL, sizeof(*usemask), PROT_READ | PROT_WRITE,
+                       MAP_ANONYMOUS | MAP_SHARED, -1, 0);
+        assert_se(usemask != NULL);
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c]: This doesn't have a side effect, so `assert(usemask);` is enough.

File: src/tty-ask-password-agent/tty-ask-password-agent.c [hunk 12]
```diff
@@ -210,6 +229,69 @@ static int ask_password_plymouth(
...
+        struct console *con = NULL;
+        size_t con_len = 0, len;
+        int ret;
+
+        assert(num);
+        assert(*num == 0);
+
+        ret = read_one_line_file("/sys/class/tty/console/active", &active);
+        if (ret < 0)
+                return con;
```
Comment (zonque) [inline src/tty-ask-password-agent/tty-ask-password-agent.c:255]: Why not just `return NULL;`?

But probably, this function should just up-chain the actual error, and return `con` as pointer, just like `num`.

Comment (bitstreamout) [inline src/tty-ask-password-agent/tty-ask-password-agent.c:255]: OK .. I'll close this pull request to be able to open a new one with a slightly changed version

**Comment by bitstreamout** [comment]
Close for this one for now ...

**Comment by karelzak** [comment]
I think use per console process is more robust solution as you want to have independent session and process group for each console (for setsid(), stdout/stdin/stderr mapping to console, etc.).

**Comment by bitstreamout** [comment]
Thanks a lot, I hope that with the pull request https://github.com/systemd/systemd/pull/1416 this should be done.  I was aware that that the former approach had done this in a more indirect way.
